### ISSUE TITLE
Bad performance due to GIL when reading groups/datasets

### ISSUE THREAD

**Issue by maartenbreddels** [h5py/h5py#1782]
Related https://github.com/h5py/h5py/issues/1516

When reading datasets, or groups, there is a lot of GIL taking and dropping, lead to bad performance in multithreaded environments. I think this is best demonstrated by code:
```python
import os
import sys
import threading

import h5py
import gil_load


try:
    gil_load.init()
    use_gil_load = True
except RuntimeError:
    # skip if not started with python -m gil_load
    use_gil_load = False


names = [f'd{i}' for i in range(20)]
filename = 'test.hdf5'


if not os.path.exists(filename):
    with h5py.File(filename, "w") as f:
        for name in names:
            f.create_dataset(name, (20,), dtype='i8')

f = h5py.File(filename)
def run():
    _ = [f[name] for name in names]

def main(argv=sys.argv):
    threads = [threading.Thread(target=run) for _ in range(int(sys.argv[1]))]
    if use_gil_load:
        gil_load.start()
    for thread in threads:
        thread.start()
    for thread in threads:
        thread.join()
    if use_gil_load:
        gil_load.stop()
        stats = gil_load.get()
        print(gil_load.format(stats))

main()
```

This can be run using `gilload`, e.g.
```
$ python -m gil_load use_case/h5pygil.py 10                                                                   dev
held: 0.81 (0.81, 0.81, 0.81)
wait: 0.952 (0.952, 0.952, 0.952)
...
```

Which means the GIL is held often, and the threads often have to wait for the GIL.

This was a good test case for a GIL visualization/performance tool I've been working on, to see if we can get a better handle on this. Installing https://github.com/maartenbreddels/per4m (it's still a rough tool).

```
$ giltracer --state-detect ./use_case/h5pygil.py 3 
...
Summary of threads:

PID         total(us)    no gil%✅    has gil%❗    gil wait%❌
--------  -----------  -----------  ------------  -------------
2793256*      59912.6         97.6           1.8            0.6
2793330        8984.5         14.0          26.7           59.3
2793331        8351.3         14.2          28.5           57.3
2793332        8130.1         13.6          27.6           58.8

High 'no gil' is good (✅), we like low 'has gil' (❗),
 and we don't want 'gil wait' (❌). (* indicates main thread)
....
```

We also see that the 3 worker threads have to wait for the GIL ~50% of the time. The VizTracer output gives us:

![image](https://user-images.githubusercontent.com/1765949/102533772-21c15800-40a6-11eb-9eff-183b82c83ef9.png)

The dark red blocks below the callstack with 'S(GIL)' in it are when a thread is in a sleeping state due to the GIL (and you see it often waking up, to check if it can continue). The light red/pink with 'GIL' cursive in it above the callstack, is when the GIL is taken. From this I think I can conclude that the h5o.open and h5i.get_type do not release the GIL.

A different view of this is a flamegraph of the callstack when the GIL is not held, and the process/thread is not running:

![image](https://user-images.githubusercontent.com/1765949/102534790-a1035b80-40a7-11eb-9c61-61a19f23d037.png)

Unfortunately I could not restore the Python callstack, but we see that ~50% of the 'offcpu, offgil time' is held by `__pyx_f_4h5py_4defs_H5Oopen` (on the right).

I wonder what the best approach here is, even if we can release the GIL, it will be taken very often on function exit again. maybe a better way to attach this issue it for instance that `group.items()` will do a single function call into cython/c without the GIL. Note that per4m/giltracer is very experimental, but I think the GIL claims are correct.

**Comment by tacaswell** [comment]
Is the problem here that we are releasing the GIL too often?  My rough understanding is that context switching between threads is not cheap but if you are about to do a "long" gil-free activity it is worth releasing it in the hope that you can let another thread do some work during that time.  If your computation takes longer than the time it takes to switch contexts you can get an effective win (because you will have many threads working at once!).  If you are the only thread releasing the GIL is mostly harmless because there is nothing else to switch to and you get it back when you return (at a small cost of the running the lock).  However, I suspect what is happening here is that because we are releasing the gil around "fast" operations but there is always another thread waiting to grab it so, we pay the high cost to switch threads, the other thread releases the GIL, does a tiny iota of work, and then we pay the cost again,...

I suspect the fix is to audit where we release the GIL and only do so in places where it is plausible that we will win the "do it in parallel" vs "pay to switch threads".

----

All of that said, there is a c++ level hdf5 global lock that needs to be acquired and a h5py global lock that needs to be acquired in addition to the GIL.  We may also be seeing lots and lots of mini-deadlocks (where even though we have the GIL we have to wait for the other thread to release the other two locks).

**Comment by takluyver** [comment]
If the problem is releasing the GIL too often, is there a good way to see which functions are releasing it?

I might make an educated guess that `H5Idec_ref` is an issue for this kind of thing. As the name suggests, this is part of HDF5's reference counting. In your code, each thread gets 20 Dataset objects and then discards them, which will lead to 20 cheap calls to `H5Idec_ref`, each of which will release and reacquire the GIL.

The challenge is that H5Idec_ref is *often* but not always cheap. Dropping the last reference to an object can trigger activity, presumably to write buffers to disk (#1521). cc @kif 

Does this have a meaningful impact on real use-cases? Accessing HDF5 files from several threads is not a priority, because as @tacaswell mentioned, HDF5 has its own global lock.

**Comment by maartenbreddels** [comment]
with a possible 100k context switches per second (https://stackoverflow.com/questions/304752/how-to-estimate-the-thread-context-switching-overhead) I don't think we are releasing the GIL too often. With a typical `h5o.open` taking 1-5ms, that overhead is really small.

So I don't think releasing too often is an issue, it's simply keeping the GIL everywhere (even in the c layer).

The best solution would be that a single `group.items()` would release the GIL once, and take the GIL again after the 100 groups are read from disk, which would require only 1 GIL take and drop.

But, the other locks sound like bad news. I think I need to resort the multiprocessing to speed this up.

In any case, I think `h5o.open` is not releasing the GIL, and I think it would be good to release it.

PS: I'm working on a tool to automate the detection of c functions that don't release the GIL, in case that might be useful in the future.

**Comment by tacaswell** [comment]
What version of h5py are you testing with @maartenbreddels ?  I think the amount of GIL releasing we did has changed recently.

**Comment by maartenbreddels** [comment]
I'm using master, cheers


(from mobile phone)

**Comment by tacaswell** [comment]
Ah, I see I am wrong about how much we drop the gil (the layer that builds the bottom most cython wrappers is controled by https://github.com/h5py/h5py/blob/master/h5py/api_functions.txt and the functions where we drop the gil are makred with `nogil` which seems to mostly be the dataset operations).  I think what this will enable is to not block computation while doing I/O.  In the cases where open is waiting for the GIL it is because the interpreter took it away from that thread (rather than the code positively dropping it)?  If that is the case, I think most of my analysis above should be disregarded, sorry for the noise.

I stand by doing a careful audit of all of our layers of locking being valuable, but likely a lot of work and the hdf5 c++ level library lock is likely to be the biggest problem here.
